docs/source/more_types.rst

```diff
@@ -178,6 +183,386 @@ create subclasses of these objects either.
+--------------------------
+
+Mypy will perform several checks on your overload variant definitions
+to ensure they behave as expected. First, mypy will check and make sure
+that no overload variant is shadowing a subsequent one. For example,
+consider the following function which adds together two ``Expression``
+objects, and contains a special-case to handle receiving two ``Literal``
+types:
+
+.. code-block:: python
+
+    from typing import overload, Union
+
+    class Expression:
+        # ...snip...
+
+    class Literal(Expression):
+        # ...snip...
+
+    # Warning -- the first overload variant shadows the second!
+
+    @overload
+    def add(left: Expression, right: Expression) -> Expression: ...
+
+    @overload
+    def add(left: Literal, right: Literal) -> Literal: ...
+
+    def add(left: Expression, right: Expression) -> Expression:
+        # ...snip...
+
+While this code snippet is technically type-safe, it does contain an
+anti-pattern: the second variant will never be selected! If we try calling
+``add(Literal(3), Literal(4))``, mypy will always pick the first variant
+and evaluate the function call to be of type ``Expression``, not ``Literal``.
+This is because ``Literal`` is a subtype of ``Expression``, which means
+the "pick the first match" rule will always halt after considering the
+first overload.
+
+Because having an overload variant that can never be matched is almost
+certainly a mistake, mypy will report an error. To fix the error, we can
+either 1) delete the second overload or 2) swap the order of the overloads:
+
+.. code-block:: python
+
+    # Everything is ok now -- the variants are correctly ordered
+    # from most to least specific.
+
+    @overload
+    def add(left: Literal, right: Literal) -> Literal: ...
+
+    @overload
+    def add(left: Expression, right: Expression) -> Expression: ...
+
+    def add(left: Expression, right: Expression) -> Expression:
+        # ...snip...
+
+Mypy will also type check the different variants and flag any overloads
+that have inherently unsafely overlapping variants. For example, consider
+the following unsafe overload definition:
+
+.. code-block:: python
+
+    from typing import overload, Union
+
+    @overload
+    def unsafe_func(x: int) -> int: ...
+
+    @overload
+    def unsafe_func(x: object) -> str: ...
+
+    def unsafe_func(x: object) -> Union[int, str]:
+        if isinstance(x, int):
+            return 42
+        else:
+            return "some string"
+
+On the surface, this function definition appears to be fine. However, it will
+result in a discrepency between the inferred type and the actual runtime type
+when we try using it like so:
+
+.. code-block:: python
+
+    some_obj: object = 42
+    unsafe_func(some_obj) + " danger danger"  # Type checks, yet crashes at runtime!
+
+This program type checks according to the annotations, but will actually crash
+at runtime! Since ``some_obj`` is of type ``object``, mypy will decide that
```

Comment:
The sentence "This program..." just repeats the comment above. I would just remove it. (Especially that we have an explanation in the next sentence.)